### PR TITLE
Using macro to reduce duplicated parse code

### PR DESCRIPTION
**Pull request by hackaugusto**
## Describe your changes

This adds a small macro to parse some of the u32 instructions. There are many other regularities in the parsing code that can be changed in the same fashion, but I'm opening this PR first to see if people consider that an improvement, prior to doing it for other opcodes.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: assembly/src/parsers/u32_ops.rs [hunk 1]
```diff
@@ -29,175 +29,123 @@ pub fn parse_u32assert(op: &Token) -> Result<Node, ParsingError> {
...
-    }
-}
+parseop!(
+    /// Returns `U32CheckedAdd` instruction node if no immediate value is provided or
+    /// `U32CheckedAddImm` instruction node otherwise.
+    ///
+    /// # Errors
+    /// Returns an error if the instruction token contains wrong number of parameters, or if the
+    /// provided parameter is not a u32 value.
+    fn parse_u32checked_add "u32checked_add" => (U32CheckedAdd, U32CheckedAddImm)
```
Comment (hackaugusto) [inline assembly/src/parsers/u32_ops.rs:57]: I experimented a bit with other syntax. This is an option that is concise and readable, but better suggestions are welcome :)

File: assembly/src/parsers/u32_ops.rs [hunk 2]
```diff
@@ -29,175 +29,123 @@ pub fn parse_u32assert(op: &Token) -> Result<Node, ParsingError> {
...
+macro_rules! parseop {
+    ($(#[$attr:meta])* fn $fn: ident $op: literal => ( $ins: ident , $imm_ins: ident)) => {
+        $(#[$attr])*
+        pub fn $fn(op: &Token) -> Result<Node, ParsingError> {
+            debug_assert_eq!(op.parts()[0], $op);
+            match op.num_parts() {
+                0 => unreachable!(),
+                1 => Ok(Instruction($ins)),
+                2 => {
+                    let value = parse_param::<u32>(op, 1)?;
```
Comment (hackaugusto) [inline assembly/src/parsers/u32_ops.rs:41]: This is my motivation for the macro, the line reduction is a nice to have, but what I really wanted was an easy way to see that all parse functions were using `u32` data type.

**Review by vlopes11 — COMMENTED** [review]
I think it's an improvement. We always need to be extra careful with macros as they are hard to debug. We should use only in cases like this where we have some boilerplate repeated code, so we reduce the chances of a copy+paste error.

The only part I think can be improved is that we can't see anywhere the function signature, given the macro definition. Even this would also be repeated code, it's always good to have the signature immediately clear just from a quick glimpse to the definition. In this case, we can't see that the function is a `fn(&Token) -> Result<Node, ParsingError>`. IMO it would be a blocker for the change, until we have some form that makes the signature clear - even tho I can't think of something immediately obvious that would be a solution.

Lacking a better option, how about naming it `parse_token_into_node_or_parsing_error`?

**Comment by bobbinth** [comment]
I'm actually of split mind about this change. On the one hand it does reduce boilerplate, but on the other I'm not sure it really improves readability (and may even make it a bit less readable as @vlopes11 mentioned).

**Comment by grjte** [comment]
> The only part I think can be improved is that we can't see anywhere the function signature, given the macro definition. Even this would also be repeated code, it's always good to have the signature immediately clear just from a quick glimpse to the definition.

I agree with this. I think losing the signature hurts readability a bit too much for my taste. If we could resolve that then I'd be in favor of the change, because I think the macro is otherwise readable and enforces more consistency by reducing duplication

**Comment by hackaugusto** [comment]
Here is how it looks with the full signature as part of the macro syntax:

```rust

macro_rules! parseop {
    (
        $(#[$attr:meta])*
        fn $fn: ident($op: literal, op: &Token) -> Result<Node, ParsingError>  => ( $ins: ident , $imm_ins: ident)
    ) => {
        $(#[$attr])*
        pub fn $fn(op: &Token) -> Result<Node, ParsingError> {
            debug_assert_eq!(op.parts()[0], $op);
            match op.num_parts() {
                0 => unreachable!(),
                1 => Ok(Instruction($ins)),
                2 => {
                    let value = parse_param::<u32>(op, 1)?;
                    Ok(Instruction($imm_ins(value)))
                }
                _ => Err(ParsingError::extra_param(op)),
            }
        }
    };
}

parseop!(
    /// Returns `U32CheckedAdd` instruction node if no immediate value is provided or
    /// `U32CheckedAddImm` instruction node otherwise.
    ///
    /// # Errors
    /// Returns an error if the instruction token contains wrong number of parameters, or if the
    /// provided parameter is not a u32 value.
    fn parse_u32checked_add("u32checked_add", op: &Token) -> Result<Node, ParsingError> => (U32CheckedAdd, U32CheckedAddImm)
);
```

Another option is to move the instruction name together with the instructions:

```rust
macro_rules! parseop {
    (
        $(#[$attr:meta])*
        fn $fn: ident(op: &Token) -> Result<Node, ParsingError> :: [$op: literal, $ins: ident , $imm_ins: ident]
    ) => {
        $(#[$attr])*
        pub fn $fn(op: &Token) -> Result<Node, ParsingError> {
            debug_assert_eq!(op.parts()[0], $op);
            match op.num_parts() {
                0 => unreachable!(),
                1 => Ok(Instruction($ins)),
                2 => {
                    let value = parse_param::<u32>(op, 1)?;
                    Ok(Instruction($imm_ins(value)))
                }
                _ => Err(ParsingError::extra_param(op)),
            }
        }
    };
}

parseop!(
    /// Returns `U32CheckedAdd` instruction node if no immediate value is provided or
    /// `U32CheckedAddImm` instruction node otherwise.
    ///
    /// # Errors
    /// Returns an error if the instruction token contains wrong number of parameters, or if the
    /// provided parameter is not a u32 value.
    fn parse_u32checked_add(op: &Token) -> Result<Node, ParsingError>
    :: ["u32checked_add", U32CheckedAdd, U32CheckedAddImm]
);
```

or with text instead of symbols:

```rust
macro_rules! parseop {
    (
        $(#[$attr:meta])*
        fn $fn: ident(op: &Token) -> Result<Node, ParsingError> for inst $op: literal with opcodes ($ins: ident , $imm_ins: ident)
    ) => {
        $(#[$attr])*
        pub fn $fn(op: &Token) -> Result<Node, ParsingError> {
            debug_assert_eq!(op.parts()[0], $op);
            match op.num_parts() {
                0 => unreachable!(),
                1 => Ok(Instruction($ins)),
                2 => {
                    let value = parse_param::<u32>(op, 1)?;
                    Ok(Instruction($imm_ins(value)))
                }
                _ => Err(ParsingError::extra_param(op)),
            }
        }
    };
}

parseop!(
    /// Returns `U32CheckedAdd` instruction node if no immediate value is provided or
    /// `U32CheckedAddImm` instruction node otherwise.
    ///
    /// # Errors
    /// Returns an error if the instruction token contains wrong number of parameters, or if the
    /// provided parameter is not a u32 value.
    fn parse_u32checked_add(op: &Token) -> Result<Node, ParsingError>
    for inst "u32checked_add" with opcodes (U32CheckedAdd, U32CheckedAddImm)
);
```

**Comment by grjte** [comment]
> or with text instead of symbols:
> 
> ```rust
> macro_rules! parseop {
>     (
>         $(#[$attr:meta])*
>         fn $fn: ident(op: &Token) -> Result<Node, ParsingError> for inst $op: literal with opcodes ($ins: ident , $imm_ins: ident)
>     ) => {
>         $(#[$attr])*
>         pub fn $fn(op: &Token) -> Result<Node, ParsingError> {
>             debug_assert_eq!(op.parts()[0], $op);
>             match op.num_parts() {
>                 0 => unreachable!(),
>                 1 => Ok(Instruction($ins)),
>                 2 => {
>                     let value = parse_param::<u32>(op, 1)?;
>                     Ok(Instruction($imm_ins(value)))
>                 }
>                 _ => Err(ParsingError::extra_param(op)),
>             }
>         }
>     };
> }
> 
> parseop!(
>     /// Returns `U32CheckedAdd` instruction node if no immediate value is provided or
>     /// `U32CheckedAddImm` instruction node otherwise.
>     ///
>     /// # Errors
>     /// Returns an error if the instruction token contains wrong number of parameters, or if the
>     /// provided parameter is not a u32 value.
>     fn parse_u32checked_add(op: &Token) -> Result<Node, ParsingError>
>     for inst "u32checked_add" with opcodes (U32CheckedAdd, U32CheckedAddImm)
> );
> ```

I quite like this last one actually. It's very readable, but we still get the benefits of putting the boilerplate in one place. Curious what others thing.

**Comment by bobbinth** [comment]
> I quite like this last one actually. It's very readable, but we still get the benefits of putting the boilerplate in one place. Curious what others thing.

I'm actually still not sure it improves things much (in terms of readability) over the original. Cognitively, there is actually a bit more I need to do here to understand what is going on. But it might be just me.

**Comment by grjte** [comment]
> I'm actually still not sure it improves things much (in terms of readability) over the original. Cognitively, there is actually a bit more I need to do here to understand what is going on. But it might be just me.

As I understand it, the goal of the change is not to improve readability. I think the goal is to encapsulate it so that there's less duplication and there's a clearer safety guarantee that similar instructions are implemented the same way. In my mind, as long as readability isn't reduced then it's an improvement

**Comment by grjte** [comment]
I haven't done a full review, since we need to decide if we want to consider this type of change. @bobbinth @vlopes11 @hackaugusto what are your thoughts on the discussion above? In my opinion, it's worth doing to reduce duplication, as long as we don't reduce readability

**Comment by vlopes11** [comment]
I'm kinda on a split decision because it reduces the code duplication, but at a small cost of readability + we create a new internal syntax in our stack. This can come at some small cost to the learning curve for people who are starting to contribute.

Since macro provides virtually free possibilities, I think it's always a good call to stick as much as possible to Rust syntax. For that reason, I think this is the best listed option

```rust
parseop!(
    /// Returns `U32CheckedAdd` instruction node if no immediate value is provided or
    /// `U32CheckedAddImm` instruction node otherwise.
    ///
    /// # Errors
    /// Returns an error if the instruction token contains wrong number of parameters, or if the
    /// provided parameter is not a u32 value.
    fn parse_u32checked_add("u32checked_add", op: &Token) -> Result<Node, ParsingError> => (U32CheckedAdd, U32CheckedAddImm)
);
```

I usually avoid using macros to define public API, but since this is solely internal, this wouldn't be a problem.

**Comment by hackaugusto** [comment]
We haven't reached a consensus, so I'm going to close this PR.

**Pull request closed without merging**